commands/run.sh

```diff
@@ -0,0 +1,542 @@
+if [[ "${BUILDKITE_PLUGIN_DOCKER_PROPAGATE_ENVIRONMENT:-false}" =~ ^(true|on|1)$ ]] ; then
+  if [[ -n "${BUILDKITE_ENV_FILE:-}" ]] ; then
+    # Read in the env file and convert to --env params for docker
+    # This is because --env-file doesn't support newlines or quotes per https://docs.docker.com/compose/env-file/#syntax-rules
+    while read -r var; do
+      args+=( --env "${var%%=*}" )
+    done < "$BUILDKITE_ENV_FILE"
+  else
+    echo -n "🚨 Not propagating environment variables to container as \$BUILDKITE_ENV_FILE is not set"
+  fi
+fi
+
+# Propagate aws auth environment variables into the container e.g. from assume role plugins
+if [[ "${BUILDKITE_PLUGIN_DOCKER_PROPAGATE_AWS_AUTH_TOKENS:-false}" =~ ^(true|on|1)$ ]] ; then
+  if [[ -n "${AWS_ACCESS_KEY_ID:-}" ]] ; then
+      args+=( --env "AWS_ACCESS_KEY_ID" )
+  fi
+  if [[ -n "${AWS_SECRET_ACCESS_KEY:-}" ]] ; then
+      args+=( --env "AWS_SECRET_ACCESS_KEY" )
+  fi
+  if [[ -n "${AWS_SESSION_TOKEN:-}" ]] ; then
+      args+=( --env "AWS_SESSION_TOKEN" )
+  fi
+  if [[ -n "${AWS_REGION:-}" ]] ; then
+      args+=( --env "AWS_REGION" )
+  fi
+  if [[ -n "${AWS_DEFAULT_REGION:-}" ]] ; then
+      args+=( --env "AWS_DEFAULT_REGION" )
+  fi
+  if [[ -n "${AWS_ROLE_ARN:-}" ]] ; then
+      args+=( --env "AWS_ROLE_ARN" )
+  fi
+  if [[ -n "${AWS_STS_REGIONAL_ENDPOINTS:-}" ]] ; then
+      args+=( --env "AWS_STS_REGIONAL_ENDPOINTS" )
+  fi
+  # Pass ECS variables when the agent is running in ECS
+  # https://docs.aws.amazon.com/sdkref/latest/guide/feature-container-credentials.html
+  if [[ -n "${AWS_CONTAINER_CREDENTIALS_FULL_URI:-}" ]] ; then
+      args+=( --env "AWS_CONTAINER_CREDENTIALS_FULL_URI" )
+  fi
+  if [[ -n "${AWS_CONTAINER_CREDENTIALS_RELATIVE_URI:-}" ]] ; then
+      args+=( --env "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI" )
+  fi
+  if [[ -n "${AWS_CONTAINER_AUTHORIZATION_TOKEN:-}" ]] ; then
+      args+=( --env "AWS_CONTAINER_AUTHORIZATION_TOKEN" )
+  fi
+  # Pass EKS variables when the agent is running in EKS
+  # https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html
+  if [[ -n "${AWS_WEB_IDENTITY_TOKEN_FILE:-}" ]] ; then
+      args+=( --env "AWS_WEB_IDENTITY_TOKEN_FILE" )
+      # Add the token file as a volume
+      args+=( --volume "${AWS_WEB_IDENTITY_TOKEN_FILE}:${AWS_WEB_IDENTITY_TOKEN_FILE}" )
+  fi
+fi
+
+if [[ "${BUILDKITE_PLUGIN_DOCKER_EXPAND_IMAGE_VARS:-false}" =~ ^(true|on|1)$ ]] ; then
+  image=$(eval echo "${BUILDKITE_PLUGIN_DOCKER_IMAGE}")
+else
+  image="${BUILDKITE_PLUGIN_DOCKER_IMAGE}"
+fi
+
+if [[ "${BUILDKITE_PLUGIN_DOCKER_ALWAYS_PULL:-false}" =~ ^(true|on|1)$ ]] ; then
+  echo "--- :docker: Pulling ${image}"
+  if ! retry "${BUILDKITE_PLUGIN_DOCKER_PULL_RETRIES:-3}" \
+       docker pull "${image}" ; then
+    echo "!!! :docker: Pull failed."
+    # retry_exit_status is set in the retry function, don't warn about it being unassigned
+    # shellcheck disable=SC2154
+    exit "$retry_exit_status"
+  fi
```

Comment:
A better way to do this without disabling shellcheck nor getting a warning/error is using the fact that the `retry` fuction actually `returns` the value to use so it is assigned to its exit code:

```suggestion
  if ! retry "${BUILDKITE_PLUGIN_DOCKER_PULL_RETRIES:-3}" \
       docker pull "${image}" ; then
    retry_exit_status="$?"
    echo "!!! :docker: Pull failed."
    exit "$retry_exit_status"
  fi
```

I tested this myself:

```
$ function test_return { local exit_code=$1; return $exit_code; }
$ test_return 1
$ echo $?
1
$ test_return 0
$ echo $?
0
$ if test_return 0; then echo "OK: $?"; else echo "FAIL: $?"; fi
OK: 0
$ if test_return 234; then echo "OK: $?"; else echo "FAIL: $?"; fi
FAIL: 234
```